### PR TITLE
provider: upgrade API Management to 2019-12-01

### PR DESCRIPTION
**Pull request by katbyte**

### CODE REVIEW

File: azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go [hunk 1]
```diff
@@ -43,8 +43,10 @@ func resourceArmApiManagementSubscription() *schema.Resource {
 				ValidateFunc: validation.Any(validation.IsUUID, validation.StringIsEmpty),
 			},
 
+			// 3.0 this seems to have been renamed to owner id?
 			"user_id": azure.SchemaApiManagementChildID(),
 
+			// 3.0 this seems to have been renamed to scope?
 			"product_id": azure.SchemaApiManagementChildID(),
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go:50]: This is not a simple rename.

An apim subscription can be tied to:
- a product -> "/products/{productId}"
- all apis -> "/apis"
- a single api -> "/apis/{apiId}"

So i think you must create a new block called "scope" with all the 3 options and in the docs you must document that only one must be filled link:

```
scope {
all_apis: boolean
api_name: string
product_name: string
}
```

used as:

scope {
product_name: "myproduct"
}

and than set the scope in SubscriptionCreateParameterProperties like
Scope: "/products/myproduct"

Hope it make sense.

File: azurerm/internal/services/apimanagement/client/client.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ type Client struct {
 	ProductApisClient          *apimanagement.ProductAPIClient
 	ProductGroupsClient        *apimanagement.ProductGroupClient
 	ProductPoliciesClient      *apimanagement.ProductPolicyClient
-	PropertyClient             *apimanagement.PropertyClient
+	NamedValueClient           *apimanagement.NamedValueClient
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/client/client.go:29]: Not ordered alphabetically

File: azurerm/internal/services/apimanagement/client/client.go [hunk 3]
```diff
@@ -95,8 +95,8 @@ func NewClient(o *common.ClientOptions) *Client {
 	productPoliciesClient := apimanagement.NewProductPolicyClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
 	o.ConfigureClient(&productPoliciesClient.Client, o.ResourceManagerAuthorizer)
 
-	propertyClient := apimanagement.NewPropertyClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
-	o.ConfigureClient(&propertyClient.Client, o.ResourceManagerAuthorizer)
+	namedValueClient := apimanagement.NewNamedValueClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/client/client.go:98]: Not ordered alphabetically

File: azurerm/internal/services/apimanagement/client/client.go [hunk 4]
```diff
@@ -134,7 +134,7 @@ func NewClient(o *common.ClientOptions) *Client {
 		ProductApisClient:          &productApisClient,
 		ProductGroupsClient:        &productGroupsClient,
 		ProductPoliciesClient:      &productPoliciesClient,
-		PropertyClient:             &propertyClient,
+		NamedValueClient:           &namedValueClient,
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/client/client.go:137]: Not ordered alphabetically

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go [hunk 5]
```diff
@@ -620,10 +621,14 @@ func resourceArmApiManagementServiceDelete(d *schema.ResourceData, meta interfac
 	name := id.Path["service"]
 
 	log.Printf("[DEBUG] Deleting API Management Service %q (Resource Grouo %q)", name, resourceGroup)
-	resp, err := client.Delete(ctx, resourceGroup, name)
+	future, err := client.Delete(ctx, resourceGroup, name)
 	if err != nil {
-		if !utils.ResponseWasNotFound(resp) {
-			return fmt.Errorf("Error deleting API Management Service %q (Resource Group %q): %+v", name, resourceGroup, err)
+		return fmt.Errorf("deleting API Management Service %q (Resource Group %q): %+v", name, resourceGroup, err)
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/resource_arm_api_management.go:626]:
```suggestion
		return fmt.Errorf("Error deleting API Management Service %q (Resource Group %q): %+v", name, resourceGroup, err)
```

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go [hunk 6]
```diff
@@ -620,10 +621,14 @@ func resourceArmApiManagementServiceDelete(d *schema.ResourceData, meta interfac
...
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("waiting for deletion of API Management Service %q (Resource Group %q): %+v", name, resourceGroup, err)
```
Comment (francescopersico) [inline azurerm/internal/services/apimanagement/resource_arm_api_management.go:631]:
```suggestion
			return fmt.Errorf("Error waiting for deletion of API Management Service %q (Resource Group %q): %+v", name, resourceGroup, err)
```

File: azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go:50]: thinking about this, this probably wants two fields here: `api_id` and `product_id` - both of which are optional & conflict with each other; which'll allow all 3 conditions to be met

File: azurerm/internal/services/apimanagement/resource_arm_api_management.go (hunk 5 above)
Comment (katbyte) [reply azurerm/internal/services/apimanagement/resource_arm_api_management.go:626]: We have decided to start removing error from wrapping messages as it becomes redundant

**Comment by katbyte** [comment]
Thanks for the review @francescopersico! really appreciate the info on how the scope changed worked. i'll see about getting this in today

File: azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go (hunk 1 above)
Comment (katbyte) [reply azurerm/internal/services/apimanagement/resource_arm_api_management_subscription.go:50]: @francescopersico - i believe as i have coded it the original behaviour will be preserved? i'll leave adding the additional properties/block to a future PR by someone who knows this service better then I

**Pull request merged**